homeassistant/components/wsdot/config_flow.py

```diff
@@ -0,0 +1,176 @@
+"""Adds config flow for wsdot."""
+
+import logging
+from types import MappingProxyType
+from typing import Any
+
+from aiohttp.client_exceptions import ClientError
+import voluptuous as vol
+import wsdot as wsdot_api
+
+from homeassistant.config_entries import (
+    SOURCE_USER,
+    ConfigEntry,
+    ConfigFlow,
+    ConfigFlowResult,
+    ConfigSubentry,
+    ConfigSubentryFlow,
+    SubentryFlowResult,
+)
+from homeassistant.const import CONF_API_KEY, CONF_ID, CONF_NAME
+from homeassistant.core import callback
+from homeassistant.helpers.selector import selector
+
+from .const import (
+    CONF_TRAVEL_TIMES,
+    DIALOG_API_KEY,
+    DIALOG_NAME,
+    DIALOG_ROUTE,
+    DOMAIN,
+    SUBENTRY_TRAVEL_TIMES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class InvalidApiKeyError(ClientError):
+    """Exception indicating the user entered an invalid API Key."""
+
+
+class WSDOTConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Config flow for WSDOT."""
+
+    VERSION = 1
+    MINOR_VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                wsdot_travel_times = wsdot_api.WsdotTravelTimes(
+                    user_input[DIALOG_API_KEY]
+                )
+                travel_time_routes = await wsdot_travel_times.get_all_travel_times()
+            except wsdot_api.WsdotTravelError as ws_error:
+                if ws_error.status == 400:
+                    errors[DIALOG_API_KEY] = (
+                        "Invalid API Key. If you do not have an API Key, you can get a new one at https://wsdot.wa.gov/traffic/api/"
+                    )
+                else:
+                    err_msg = "Unable to retrieve routes from WSDOT"
+                    _LOGGER.exception(err_msg)
+                    errors["base"] = err_msg
+            else:
+                return self.async_create_entry(
+                    title=user_input[DIALOG_NAME],
+                    data={
+                        CONF_API_KEY: user_input[DIALOG_API_KEY],
+                        CONF_TRAVEL_TIMES: [
+                            {"id": t.TravelTimeID, "name": t.Name}
+                            for t in travel_time_routes
+                        ],
```

Comment:
what's in this data?

Reply:
I am just a little confused as to why we store all the travel times in the entry data

Reply:
That's better as routes can change. I am all for minimalizing the amount of requests we do to fetch this, but I don't think writing to disk is the solution here